src/Features/CSharp/Portable/ForeachToFor/CSharpForeachToForCodeRefactoringProvider.cs

```diff
@@ -0,0 +1,160 @@
+using Microsoft.CodeAnalysis.Simplification;
+using Microsoft.CodeAnalysis.Text;
+
+namespace Microsoft.CodeAnalysis.CSharp.ForeachToFor
+{
+    [ExportCodeRefactoringProvider(LanguageNames.CSharp, Name = nameof(CSharpForeachToForCodeRefactoringProvider)), Shared]
+    internal sealed class CSharpForeachToForCodeRefactoringProvider : AbstractForeachToForCodeRefactoringProvider
+    {
+        public CSharpForeachToForCodeRefactoringProvider()
+        {
+        }
+
+        protected override SyntaxNode GetForeachStatement(SyntaxToken token)
+        {
+            var foreachStatement = token.Parent.FirstAncestorOrSelf<ForEachStatementSyntax>();
+            if (foreachStatement == null)
+            {
+                return null;
+            }
+
+            // support refactoring only if caret is in between "foreach" and ")"
+            var scope = TextSpan.FromBounds(foreachStatement.ForEachKeyword.Span.Start, foreachStatement.CloseParenToken.Span.End);
+            if (!scope.IntersectsWith(token.Span))
+            {
+                return null;
+            }
+
+            // check whether there is any comments between foreach and ) tokens
+            // if they do, we don't support conversion.
+            foreach (var trivia in foreachStatement.DescendantTrivia(n => n == foreachStatement || scope.Contains(n.FullSpan)))
+            {
+                if (trivia.Span.End <= scope.Start ||
+                    scope.End <= trivia.Span.Start)
+                {
+                    continue;
+                }
+
+                if (trivia.Kind() != SyntaxKind.WhitespaceTrivia &&
+                    trivia.Kind() != SyntaxKind.EndOfLineTrivia)
+                {
+                    // we don't know what to do with these comments
+                    return null;
+                }
+            }
+
+            return foreachStatement;
+        }
+
+        protected override (SyntaxNode start, SyntaxNode end) GetForeachBody(SyntaxNode node)
+        {
+            var foreachStatement = (ForEachStatementSyntax)node;
+            return (foreachStatement.Statement, foreachStatement.Statement);
+        }
+
+        protected override void ConvertToForStatement(SemanticModel model, ForeachInfo foreachInfo, SyntaxEditor editor, CancellationToken cancellationToken)
+        {
+            cancellationToken.ThrowIfCancellationRequested();
+
+            var parseOption = model.SyntaxTree.Options;
+            var foreachStatement = (ForEachStatementSyntax)foreachInfo.ForeachStatement;
+
+            // first, see whether we need to introduce new statement to capture collection
+            var foreachCollectionExpression = GetForeachCollection(foreachStatement);
+            var collectionVariableName = foreachCollectionExpression.ToString();
+            if (foreachInfo.RequireCollectionStatement)
+            {
+                collectionVariableName = CreateUniqueMethodName(model, foreachStatement, "list");
```

Comment:
the name should likely be based on the type of the item we were foreaching.  So "list" or "array".

Reply:
well. that can be improved later. I will rename it as collection.

Reply:
Please do not.  If you want to actually add this code to the IDE then we should be doing teh appropriate level of work to create features that fit what users will want.  We should name these "list" and "array" as appropriate.

Reply:
do we have code that return best name based on context?